### PR TITLE
fix bug about `CREATE ... AS WITH RECURSIVE`

### LINKED ISSUE #69659: Segfault when creating view with recursive
Hi,

I was playing around with new recursive feature and ran into a bug that crashes ClickHouse.

I was using ClickHouse version 23.8.4.13 and tried to create View like this
```sql
CREATE VIEW IF NOT EXISTS test_query AS (
  WITH RECURSIVE
    42 as ttt
  select ttt
);
```
but got segmentation fault error.

The query itself works fine tho.

It reproduces on fiddle.clickhouse

https://fiddle.clickhouse.com/9a2754d0-1b13-4cdf-9864-6e1ee65f48b3

### PR DESCRIPTION
**Pull request by hanfei1991**
fix https://github.com/ClickHouse/ClickHouse/issues/69659

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
fix crash when executing `create view t AS (WITH RECURSIVE 42 as ttt SELECT ttt);`

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/Interpreters/AddDefaultDatabaseVisitor.h [hunk 1]
```diff
@@ -121,7 +121,10 @@ class AddDefaultDatabaseVisitor
     {
         if (select.recursive_with)
             for (const auto & child : select.with()->children)
-                with_aliases.insert(child->as<ASTWithElement>()->name);
+            {
+                if (typeid_cast<ASTWithElement *>(child.get()))
```
Comment (hanfei1991) [inline src/Interpreters/AddDefaultDatabaseVisitor.h:125]: We cannot assume it has to be `ASTWithElement`, like `42 as ttt` is a string literal

File: tests/queries/0_stateless/03237_create_table_select_as_with_recursive.sql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+drop table if exists t;
+create view t AS (WITH RECURSIVE 42 as ttt SELECT ttt);
```
Comment (novikd) [inline tests/queries/0_stateless/03237_create_table_select_as_with_recursive.sql:2]: `RECURSIVE 42 as ttt` what does it even mean? I suspect we should not allow such queries.

Comment (hanfei1991) [reply tests/queries/0_stateless/03237_create_table_select_as_with_recursive.sql:2]: yes we shouldn't allow expression as identifier in `with recursive` clause. will change the parser

Comment (denya133) [reply tests/queries/0_stateless/03237_create_table_select_as_with_recursive.sql]: it's example only. 
real query might be more complex
but only one CTE in whole with declaration is recursive and other CTEs are regular.
for example

```sql
create view t AS (
  WITH RECURSIVE
    toDate({startDate:String}) AS _startDate
    , toDate({endDate:String}) AS _endDate
    , {preferredCategories:Array(UInt64)} AS _preferredCategories
    , {preferredSubjects:Array(Tuple(UInt64, UInt64))} AS _preferredSubjects
    , _source_categories AS (
      SELECT
        categoryId
        , argMax(dc.name, dc.date) AS _name
        , argMax(dc.path, dc.date) AS _path
        , argMax(dc.url, dc.date) AS _url
        , argMax(dc.parentId, dc.date) AS _parentId
      FROM daily_categories dc
      WHERE
        dc.date >= _startDate
        AND dc.date <= _endDate
      GROUP BY categoryId
    )
    , _low_level_categories AS (
        SELECT 
          _sc.*
          , [_sc.categoryId] AS path
          , []::Array(UInt64) AS parents
        FROM _source_categories _sc
        WHERE  
          _sc._parentId = 0
      UNION ALL
        SELECT 
          _sc.*
          , arrayConcat(_llc.path, [_sc.categoryId]) AS path
          , arrayConcat(_llc.path) AS parents
        FROM _source_categories _sc, _low_level_categories _llc
        WHERE 
          _sc._parentId = _llc.categoryId
    )
  SELECT 
    distinct(categoryId)
    , _name
    , _path
    , _url
    , _parentId
  FROM (
    SELECT 
      * 
    FROM ( 
      SELECT 
        *
        , groupUniqArrayArray(parents) OVER() AS _parents
      FROM _low_level_categories
    ) _llc
    WHERE
      _llc._parentId != 0
      AND not has(_parents, _llc.categoryId) 
      AND ((
        length(_preferredCategories) = 0 AND length(_preferredSubjects) = 0
      ) OR (
        length(_preferredCategories) != 0 AND hasAny(_llc.path, _preferredCategories)
      ))
  )
  SETTINGS
    allow_experimental_analyzer = 1
);
```

recursive needs for _low_level_categories only in that WITH declaration.

I guss that needs to test more complex queries too.

Comment (hanfei1991) [reply tests/queries/0_stateless/03237_create_table_select_as_with_recursive.sql]: @novikd Looks like it's a reasonable SQL, `WITH RECUSIVE (expression as alias) , (sql as table_name) ...`. What do you think?

Comment (novikd) [reply tests/queries/0_stateless/03237_create_table_select_as_with_recursive.sql]: Yes, it makes sense. I initially thought `RECURSIVE` must be a part of the CTE declaration. I agree that it is worth adding more complex test queries.

Comment (hanfei1991) [reply tests/queries/0_stateless/03237_create_table_select_as_with_recursive.sql]: @novikd updated tests, PTAL

**Review by novikd — APPROVED** [review]

**Pull request merged**